### PR TITLE
[Merged by Bors] - feat(Logic/Function): define `Pi.map`

### PR DESCRIPTION
**Pull request by urkud**
Also use it instead of `dcomp` or an inline lambda here and there.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Prod.lean [hunk 1]
```diff
@@ -822,9 +813,19 @@ theorem dcomp_image_pi {f : ∀ i, α i → β i} (hf : ∀ i ∉ s, Surjective
   choose a hab hat using this
   exact ⟨a, hat, funext hab⟩
 
-theorem dcomp_image_univ_pi (f : ∀ i, α i → β i) (t : ∀ i, Set (α i)) :
-    (f _ ∘' ·) '' univ.pi t = univ.pi fun i ↦ f i '' t i :=
-  dcomp_image_pi (by simp) t
+@[deprecated (since := "2024-10-06")] alias dcomp_image_pi := piMap_image_pi
```
Comment (grunweg) [inline Mathlib/Data/Set/Prod.lean:816]: Flagging this for other reviewers: you changed the statement of the lemma, renamed it and are adding an alias.
I think this is more helpful than harmful, but seems worth pointing out.

**Review by grunweg — APPROVED** [review]
Looks reasonable, thanks!

File: Mathlib/Data/Set/Prod.lean (hunk 1 above)
Comment (ADedecker) [reply Mathlib/Data/Set/Prod.lean:816]: No, I think it's just the GitHub view that is confusing.

Comment (urkud) [reply Mathlib/Data/Set/Prod.lean:816]: I changed the statement from `dcomp` to `Pi.map`. They are defeq but not reducibly defeq.

Comment (ADedecker) [reply Mathlib/Data/Set/Prod.lean:816]: Oh yeah ok I just misinterpreted that comment I think.

**Review by ADedecker — APPROVED** [review]
I like this change a lot, but it leaves a question: what is the use for `dcomp` now? Is the point just to have another reducibility setting? Should we deprecate `dcomp` as well?

**Comment by urkud** [comment]
In some cases `dcomp` is used for reindexing of a dependent function.

**Comment by ADedecker** [comment]
:facepalm: I should think twice before writing, of course `dcomp` is still more general. So everything is good.

bors r+

Thanks!

**Pull request closed without merging**